docs: clean up quickstart docs (#2214)
namehash/ensnode: docs/ensnode.io/src/content/docs/docs/integrate/omnigraph/concepts.mdx

---
title: ENS Omnigraph Core Concepts
description: The core data-model concepts behind the ENS Omnigraph API — Namegraph, Unigraph, canonicality, polymorphism, Interpreted Names, Events, and Permissions.
sidebar:
  label: Core Concepts
---

import { CardGrid, LinkCard } from "@astrojs/starlight/components";

> The [ENS Omnigraph API](/docs/integrate/omnigraph) abstracts away most of ENS's protocol complexity, but a handful of concepts explain _how_ it presents ENSv1 and ENSv2 through a single unified schema. You don't need these to send your first query, only to understand how the Omnigraph represents the state of the ENS protocol.

Reviewer: ```suggestion
The [ENS Omnigraph API](/docs/integrate/omnigraph) abstracts away most of the complexity of the ENS protocol. There are still some core concepts that are useful to learn. These  concepts will help you understand _how_ the Omnigraph presents ENSv1 and ENSv2 through a single unified schema. You don't need to master these concepts to build your first Omnigraph query, but as you level-up to building more advanced ENS features in your app, these concepts will help you to understand how the Omnigraph represents the state of the ENS protocol.
```

Author: Applied.

Because ENSv1 names do not actually have on-chain Subregistries, the Unigraph represents this relationship with an `ENSv1VirtualRegistry`.

```mermaid
flowchart TD
    root([ENSv1 Root Registry])
    eth["eth"]
    ethReg([".eth (virtual) Subregistry"])
    vitalik["vitalik.eth"]
    exampleName["example.eth"]
    vitalikReg(["vitalik.eth's (virtual) Subregistry"])
    blog["blog.vitalik.eth"]
    exampleReg(["example.eth's (virtual) Subregistry"])
    sub["sub.example.eth"]

    root --> eth
    eth --> ethReg
    ethReg --> vitalik
    ethReg --> exampleName
    vitalik --> vitalikReg
    vitalikReg --> blog
    exampleName --> exampleReg
    exampleReg --> sub
```

You navigate the graph by following a Domain to its Subregistry to its child Domains, and so on:

```graphql title="example.gql"
query Namegraph {
  # reference a Domain by name
  domain(by: { name: "eth" }) {
    # this is the Registry that "eth" exists within

>     registry { id contract { chainId address } }

Author: `AccountId` is an object type in the Omnigraph schema (`type AccountId { address: Address!, chainId: ChainId! }`), not a scalar — so selecting `contract { chainId address }` is valid GraphQL. Resolving as a false positive.

Within the Omnigraph API the complexity of the Namegraph is reduced, and all Canonical Domains are queryable, searchable, and addressable by said Canonical Name. Domains that are not canonical are _still_ referenceable by `id` (eg. `domain(by: { id: DomainId! })`).

Canonical Domains have a `Domain.canonical` field hosting the canonicality-derived fields such as `name`, `node`, `depth` (i.e. 2 for `vitalik.eth`), and `path` from the ENS root.

```graphql title="example.gql"
query Canonicality {
  domain(by: { name: "vitalik.eth" }) {
    canonical {
      name {
        interpreted # the InterpretedName
        beautified  # the ENSIP-15 BeautifiedName
      }
      node   # namehash(name)
      depth  # i.e. 2 for "vitalik.eth"

>       path   # [Domain("eth"), Domain("vitalik.eth")]

Author: Fixed — added a sub-selection (path { id }) so the example validates.

```graphql title="example.gql"
query InterpretedNames {
  domain(by: { name: "vitalik.eth" }) {
    label {
      interpreted   # "vitalik"
      beautified    # "vitalik"
      hash          # 0xaf2caa1c2ca1d027f1ac823b529d0a67cd144264b2789fa2ea4d63a67c7103cc
    }

    canonical {
      name {
        interpreted # "vitalik.eth"
        beautified  # "vitalik.eth"
      }
    }
  }
}
```

```tsx title="Example.tsx"
<p>the Label for this domain is {domain.label.beautified}</p>
```

<CardGrid>
  <LinkCard
    title="Interpreted Name"
    description="The full definition in terminology."
    href="/docs/reference/terminology#interpreted-name"
  />
  <LinkCard
    title="Interpreted Label"
    description="The full definition in terminology."
    href="/docs/reference/terminology#interpreted-label"
  />
  <LinkCard
    title="Beautified Name"
    description="The full definition in terminology."
    href="/docs/reference/terminology#beautified-name"
  />
  <LinkCard
    title="Beautified Label"
    description="The full definition in terminology."
    href="/docs/reference/terminology#beautified-label"
  />
</CardGrid>

### InterpretedName with Encoded LabelHash Example

> As noted above, an InterpretedName may contain Labels that are [Encoded LabelHashes](/docs/reference/terminology#encoded-labelhash), meaning that the human-readable Label isn't known or isn't normalizable. The Omnigraph still supports referencing Domains by these InterpretedNames, like so:

Reviewer: ```suggestion
As noted above, an InterpretedName may contain Labels that are [Encoded LabelHashes](/docs/reference/terminology#encoded-labelhash), meaning that the human-readable Label isn't known or isn't normalized. The Omnigraph still supports referencing Domains by these InterpretedNames, like so:
```

Author: Applied.

```graphql title="example.gql"
query ByInterpretedName {
  domain(by: { name: "[af2caa1c2ca1d027f1ac823b529d0a67cd144264b2789fa2ea4d63a67c7103cc].eth" }) {
    id  # this is the same Domain as "vitalik.eth" above!
    canonical {
      name {
        # in this case, the Omnigraph knows the fully healed Canonical Name of "vitalik.eth"
        interpreted # "vitalik.eth"
      }
    }
  }
}
```

:::danger[Not all InterpretedNames are Resolvable]
ENS Forward Resolution does _not_ support Encoded LabelHashes in names, so while an InterpretedName can be used to traverse the Namegraph (i.e. with `domain(by: { name: "" })`), if the Canonical Name is not a [ResolvableName](/docs/reference/terminology#resolvable-name), then the records for that name _cannot_ be determined.
:::

### BeautifiedNames and BeautifiedLabels

> For some labels and names, there exists an [ENSIP-15](https://docs.ens.domains/ensip/15) beautified form, where certain normalized characters are 'beautified' into their _un-normalized_ but prettier forms.

Reviewer: ```suggestion
For some labels and names, there exists an [ENSIP-15](https://docs.ens.domains/ensip/15) beautified form, where certain normalized characters are 'beautified' into their _un-normalized_ but _normalizable_ prettier forms.
```

Author: Applied.

> For example, the InterpretedName `♾.eth` (normalized) would be beautified to `♾️.eth` (un-normalized). All name and label fields in the Omnigraph provide a `beautified` variant which can be used for display.

Reviewer: ```suggestion
For example, the InterpretedName `♾.eth` (normalized) would be beautified to `♾️.eth` (un-normalized, but is normalizable back to `♾.eth`). All name and label fields in the Omnigraph provide a `beautified` variant which can be used for display.
```

Author: Applied.

Because ENSv1 names do not actually have on-chain Subregistries, the Unigraph represents this relationship with an `ENSv1VirtualRegistry`.

```mermaid
flowchart TD
    root([ENSv1 Root Registry])
    eth["eth"]
    ethReg([".eth (virtual) Subregistry"])
    vitalik["vitalik.eth"]
    exampleName["example.eth"]
    vitalikReg(["vitalik.eth's (virtual) Subregistry"])
    blog["blog.vitalik.eth"]
    exampleReg(["example.eth's (virtual) Subregistry"])
    sub["sub.example.eth"]

    root --> eth
    eth --> ethReg
    ethReg --> vitalik
    ethReg --> exampleName
    vitalik --> vitalikReg
    vitalikReg --> blog
    exampleName --> exampleReg
    exampleReg --> sub
```

You navigate the graph by following a Domain to its Subregistry to its child Domains, and so on:

```graphql title="example.gql"
query Namegraph {
  # reference a Domain by name
  domain(by: { name: "eth" }) {
    # this is the Registry that "eth" exists within
    registry { id contract { chainId address } }

    # "eth"'s parent Domain (if any)
    parent { id }

    # the Subregistry that "eth" declares
    subregistry {
      domains {
        edges {
          node {

>             # get each domain's (beautified) name

Author: These are static illustrative snippets, not live-executed against the active schema. We are intentionally documenting v1.15.1 fields (e.g. `beautified`) here. The `ACTIVE_OMNIGRAPH_VERSION` bump to v1.15.1 is tracked in #2223 and lands soon; this PR also won't merge before blue is promoted to production. Deferred to #2223.

---
title: ENS Omnigraph Core Concepts
description: The core data-model concepts behind the ENS Omnigraph API — Namegraph, Unigraph, canonicality, polymorphism, Interpreted Names, Events, and Permissions.
sidebar:
  label: Core Concepts
---

import { CardGrid, LinkCard, Aside } from "@astrojs/starlight/components";

> The [ENS Omnigraph API](/docs/integrate/omnigraph) abstracts away most of the complexity of the ENS protocol. There are still some core concepts that are useful to learn. These concepts will help you understand _how_ the Omnigraph presents ENSv1 and ENSv2 through a single unified schema. You don't need to master these concepts to build your first Omnigraph query, but as you level-up to building more advanced ENS features in your app, these concepts will help you to understand how the Omnigraph represents the state of the ENS protocol.

Reviewer: ```suggestion
While the [ENS Omnigraph API](/docs/integrate/omnigraph) abstracts away most of the complexity of the ENS protocol, there are still some core concepts that are useful to learn. These concepts will help you understand _how_ the Omnigraph presents ENSv1 and ENSv2 through a single unified schema. You don't need to master these concepts to build your first Omnigraph query, but as you level-up to building more advanced ENS features in your app, these concepts will help you to understand how the Omnigraph represents the state of the ENS protocol.
```

Author: Applied.
